### PR TITLE
feat(ingest): redshift - Redshift rework

### PR DESCRIPTION
**Pull request by treff7es**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by treff7es** [comment]
CleanShot 2023-01-26 at 18 05 26@2x
CleanShot 2023-01-26 at 18 05 18@2x
CleanShot 2023-01-26 at 18 05 08@2x
CleanShot 2023-01-26 at 18 04 49@2x
CleanShot 2023-01-26 at 18 04 38@2x
CleanShot 2023-01-26 at 18 04 30@2x
CleanShot 2023-01-26 at 18 04 17@2x
CleanShot 2023-01-26 at 18 04 05@2x

**Review by hsheth2 — DISMISSED** [review]
Took a cursory look over this so far

**Comment by treff7es** [comment]
I added complex type support as well:
CleanShot 2023-02-06 at 10 35 47@2x

File: metadata-ingestion/setup.py [hunk 1]
```diff
@@ -532,8 +539,10 @@ def get_long_description():
         "oracle = datahub.ingestion.source.sql.oracle:OracleSource",
         "postgres = datahub.ingestion.source.sql.postgres:PostgresSource",
         "redash = datahub.ingestion.source.redash:RedashSource",
-        "redshift = datahub.ingestion.source.sql.redshift:RedshiftSource",
-        "redshift-usage = datahub.ingestion.source.usage.redshift_usage:RedshiftUsageSource",
```
Comment (anshbansal) [inline metadata-ingestion/setup.py:536]: breaking change. Please see if we can avoid it. Also, need to update [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md) file

File: metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+                    scan_type in (1, 2, 3)
+                order by cluster, target_schema, target_table, starttime asc
+            """.format(
+            # We need the original database name for filtering
+            db_name=self.config.database,
+            start_time=self.config.start_time.strftime(redshift_datetime_format),
+            end_time=self.config.end_time.strftime(redshift_datetime_format),
+        )
+        view_lineage_query = """
+                select
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py:352]: can we locate these inside a queries.py constants file?

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py [hunk 3]
```diff
@@ -256,62 +230,88 @@ def _gen_operation_aspect_workunits(
...
-        if not self.config.table_pattern.allowed(full_table_name):
-            logger.debug(f"Filtering out {full_table_name} due to table_pattern.")
-            self.report.report_dropped(full_table_name)
+        if (
+            event.database not in all_tables
+            or event.schema_ not in all_tables[event.database]
+            or not any(
+                event.table == t.name for t in all_tables[event.database][event.schema_]
+            )
+        ):
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:254]: Can just return `not` this expression, or the demorgan's law equivalent

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py [hunk 4]
```diff
@@ -256,62 +230,88 @@ def _gen_operation_aspect_workunits(
...
+                    )
+                    self.report.num_usage_stat_skipped += 1
+                    continue
+
+                # Replace database name with the alias name if one is provided in the config.
+                if self.config.database_alias:
+                    access_event.database = self.config.database_alias
+
+                if not self._should_process_event(access_event, all_tables=all_tables):
+                    # TODO: Add to report number of skipped event
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:301]: Is this completed?

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py [hunk 5]
```diff
@@ -256,62 +230,88 @@ def _gen_operation_aspect_workunits(
...
+                    self.report.num_usage_stat_skipped += 1
+                    continue
+
+                # yield access_event
+                access_events.append(access_event)
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:306]: Is this purposefully not yielding to fetch from the db all at once? The comment and method name make it sound like it should be a generator

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py [hunk 6]
```diff
@@ -325,6 +325,10 @@ def _gen_operation_aspect_workunits_from_access_events(
             ):
                 continue
 
+            if not self._should_process_event(event, all_tables=all_tables):
+                self.report.num_operational_stats_skipped += 1
+                continue
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:330]: Think is checked twice, once here and once in `_gen_access_events_from_history_query`

**Review by asikowitz — COMMENTED** [review]
Sorry for the incredibly long review. Comments range from typos and very minor style suggestions, to larger refactoring suggestions, to plain questions. Kinda glad I'm picking this up because this is an overwhelming number of comments

File: metadata-ingestion/src/datahub/ingestion/source/redshift/redshift.py [hunk 7]
```diff
@@ -0,0 +1,919 @@
...
+                lineage_info = self.lineage_extractor.get_lineage(
+                    table,
+                    dataset_urn,
+                    self.db_schemas[database][schema],
+                )
+                if lineage_info:
+                    for wu in gen_lineage(
+                        dataset_urn, lineage_info, self.config.incremental_lineage
+                    ):
+                        yield wu
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/redshift.py:899]: Can use a `yield from`; same below

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py [hunk 8]
```diff
@@ -256,62 +229,81 @@ def _gen_operation_aspect_workunits(
...
+                        starttime=row[field_names.index("starttime")],
+                        endtime=row[field_names.index("endtime")],
+                        operation_type=row[field_names.index("operation_type")]
+                        if "operation_type" in field_names
+                        else None,
+                    )
+                except pydantic.error_wrappers.ValidationError as e:
+                    logging.debug(
+                        f"Validation error on access event creation from row {row}. The error was: {e} Skipping ...."
+                    )
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:288]: Is this very common? Seems like something we should report in non-debug mode

**Review by asikowitz — COMMENTED** [review]
Really nice job on refactoring some of the more confusing code. I have a few more comments, but we should bias towards getting this in when it's in a good state over doing more style changes. Holding off approval until a merge is completed because there are some potentially breaking changes, e.g. around merging the sub type changes

**Review by asikowitz — APPROVED** [review]
LGTM! Thanks for working through the many requested changes

File: metadata-ingestion/src/datahub/ingestion/source/redshift/redshift_schema.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+        # This query needs to run separately as we can't join witht the main query because it works with
+        # driver only functions.
+        enriched_table = RedshiftDataDictionary.enrich_tables(conn)
+
+        cur = RedshiftDataDictionary.get_query_result(conn, RedshiftQuery.list_tables)
+        field_names = [i[0] for i in cur.description]
+        db_tables = cur.fetchall()
+
+        for table in db_tables:
+            schema = table[field_names.index("schema")]
```
[the remark this answers is not in the document]
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/redshift/redshift_schema.py:175]: These fields are listed in the SQL query, so they should be available. The only issue can be if you delete/rename in the query but I think we should fail in that case

File: metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py [hunk 10]
```diff
@@ -209,36 +210,11 @@ def _populate_lineage_map(
...
-                        continue
-                else:
-                    target_platform = LineageDatasetPlatform.REDSHIFT
-                    target_path = f"{alias_db_name}.{lineage_row.target_schema}.{lineage_row.target_table}"
-
-                target = LineageItem(
-                    dataset=LineageDataset(platform=target_platform, path=target_path),
-                    upstreams=set(),
-                    collector_type=lineage_type,
+                target = self._get_target_lineage(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py:213]: much nicer!

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py [hunk 11]
```diff
@@ -337,24 +337,17 @@ def _gen_operation_aspect_workunits_from_access_events(
...
-                id=f"operation-aspect-{event.table}-{event.endtime.isoformat()}",
-                mcp=mcp,
+            dataset_urn = DatasetUrn.create_from_ids(
+                platform_id="redshift",
+                table_name=resource.lower(),
+                platform_instance=self.config.platform_instance,
+                env=self.config.env,
             )
+
+            yield MetadataChangeProposalWrapper(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:347]: nice!

File: metadata-ingestion/src/datahub/specific/dataset.py [hunk 12]
```diff
@@ -113,7 +114,7 @@ def add_upstream_lineage(self, upstream: Upstream) -> "DatasetPatchBuilder":
         self._add_patch(
             "upstreamLineage",
             "add",
-            path=f"/upstreams/{upstream.dataset}",
+            path=f"/upstreams/{quote(upstream.dataset, safe='')}",
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/specific/dataset.py:117]: @hsheth2 This breaks integration tests as well but I had to do it as `s3` path with `/` caused lineage to break.

Comment (hsheth2) [reply metadata-ingestion/src/datahub/specific/dataset.py:117]: Can we change this to exclusively quote `/` characters but not need to quote everything?

Also did you test how this works with our actual patch applier?

**Comment by hsheth2** [comment]
@treff7es looks like the docs site still doesn't like the redshift-usage-legacy fixes

```
Failed to process redshift-usage-legacy due to exception No module named 'datahub.ingestion.source.usage.redshift_usage'
```

**Pull request merged**